### ISSUE TITLE
Data processing revamp design

### ISSUE THREAD

**Issue by Plyb** [DRAGNLabs/MIRROR-Pipeline#337]
**Acceptance Criteria**:
- We have a design for a data processing pipeline that supports the following criteria:
1. Ability to mix-and-match datasets and models
2. Option to not scan/process the whole dataset up front (for quick iteration)
3. Option to preprocess/cache the data (for faster repeated runs)
4. Ability to create multiple training examples from a single example or vice versa (such as an example for every MCQA option or collating multiple text examples into a longer text example)
5. Indication of progress through an epoch
6. Ability to distribute the dataset across ranks
7. Ability to shuffle per-epoch
8. mix-and-matchable multi-step preprocessing
9. Ability to avoid materializing the whole dataset in memory at the same time
10. Works with jsonargparse configs
11. Type safe as much as possible
12. Happy path is easy
13. Multiple datasets can be combined into one (such as for the mixed dataset)

# Report

Please review this deeply, since this is important! If you have questions, please let me know so I can clarify.

"take" == "head"

## Summary

| Criterion | Current | Proposal |
| ----------| ---------| ----------|
| Mix and match | ✅ | ✅ |
| No scan | ⚠️If you try doing something like a map or filter in the dataset itself, our current system will scan. Can be avoided by doing a select first (which can get messy for sub-dataset scenarios). | ⚠️ still relies on the user to be careful to not scan, but generalizes better to sub-dataset scenarios |
| Cache | ⚠️Relies on hf datasets caching system, which has some unintuitive limitations (imported functions are fingerprinted by location, not value; datasets must be file-backed to be cached; torch-format and device is inconsistent). | ⚠️ Still relies on hf, but in a way that is directly under the user's control |
| Many-1 and 1-Many | ❌ | ✅ |
| Progress Indicator | ✅ | ✅ |
| Distributed Support | ✅ | ✅ |
| Per-epoch shuffle | ✅ | ✅ |
| Multistep support | ⚠️sort of through constructor parameters. No built in way to cache intermediate steps. | ✅ |
| Partial Materialization | ✅ | ⚠️It's now up to the user to manage this, but straightforward for the happy path |
| jsonargparse compat | ✅ | ✅ |
| Type safety | ⚠️Exposes the raw, unprocessed hf dataset without type guarding. Relies on the consumer to combine that properly with `to_row_type` | ✅ |
| Happy path | ✅ | ✅ |
| Multi-sub-dataset | ❌ | ✅ |
| Notes |  |  |

## Design

There are three big, new components/changes:
1. TypedDataset[RowT]: this is a thin wrapper around a huggingface dataset that has better typing (it keeps track of its rows types)
2. `DataSource[RowT]` a factory for `TypedDataset[RowT]`s. Effectively replaces the current `MirrorDataset[RowT]` plus preprocessing steps.
3. Model-specific `DataSource` and collator constructors. These handle the mix-and-matching of datasets and models.

A big architectural change is that all preprocessing except the final tokenization and collation is handled directly by the user, giving much more flexibility.
- this means we can get rid of the preprocess function, and the `OnDemandPreprocessedDataset`

### High level algorithm

`DataSource`s can both define a leaf-node dataset (something like wikitext, or MMLU), or, if they take other `DataSource`s as constructor arguments, can define transformations on datasets (for instance, splitting training examples, or mixing two underlying datasets). Instead of having a raw, unprocessed, backing hf dataset, they have a backing `TypedDataset` that already has all of the transformations applied.
- **Why is having the already processed `TypedDataset` better than what we have now?** A few reasons:
  - In some cases, there was no logical unprocessed backing dataset, such as for `MixedDataset`
  - Better type safety, since the consumer doesn't need to be in charge of combining the unprocessed dataset with `to_row_type`
  - Datasets whose lengths have changed (through a one-to-many or many-to-one transformation) are materialized and can be handed directly to tqdm for a progress bar. In other words, it enables many-to-one and one-to-many transformations
- **Doesn't this cause a scanning problem if you're materializing the `TypedDataset`?** Yes, but if the user is careful to not ever do operations such as `map` on the full `dataset` (unless they actually need the full dataset), it should be fine. A common pattern I imagine will be having a `take` constructor parameter that is applied `before` any mapping or other scanning operations.

Once the full user-defined `DataSource` tree is constructed, it is combined with an optional formatter which is gotten from the model. This function wraps the `DataSource` in one last transformation layer that can tokenize or do any other formatting needed to get the data into a shape the model can handle. Finally, the model also optionally provides a collator to take the formatted rows and put them into a collated format, including converting to torch format and putting them on the correct device (before this point, they should be in raw python format for easy serialization with huggingface).

### Detailed types

`TypedDataset[RowT]` has a similar interface to huggingface's `Dataset`, but with strong typing. So its `__getitem__` method for instance returns a `RowT` (or slice of them if a slice index is used). `.map` returns a new `TypedDataset` with the row type defined by the function passed to `.map`.

```py
class DataSource[RowT](ABC):
    @property
    @abstractmethod
    def dataset(self) -> TypedDataset[RowT]: ...

class MirrorModel[RawT, FormattedT, BatchT]:
    def format_data(self, data_source: DataSource[RawT]) -> DataSource[FormattedT]: ... # could have a no-op default
    def collate(self, formatted_rows: list[FormattedT]) -> BatchT: ... # could have a no-op default
```

## Criteria Satisfaction

How does this satisfy the criteria?
1. Mix-and-match models and datasets: the `format_data` and `collate` functions on a model
2. No scan: must be handled by the user, but simple enough using `TypedDataset` to just `take` first before applying transformations
3. Option to preprocess/cache the data: two ways to handle this. Either just rely on hf datasets built in caching by doing `.map`, or, in cases where that fails, do it manually yourself. The happy path is still easy, but with better flexibility for edge cases. If any manual cases become common, we can write utility functions to help.
4. Many-1 and 1-many: because preprocessing is now intentionally handled through sub-datasource constructor parameters, this is straightforward to do. Each parent datasource has access to all of the items of the sub-datasource, allowing many-to-1 transformations, and the parent datasource constructs an independent `TypedDataset`, which can easy enable 1-to-many transformations.
5. Progress indication: natural, by getting the length of the constructed `TypedDataset`
6. Ability to distribute the dataset across ranks: handled by fabric, since a `TypedDataset` is a dataset and can easily use a distributed sampler (note: to get fabric to recognize that it is a map-style dataset, we might need to pass in the underlying hf dataset, but that shouldn't be an issue)
7. Ability to shuffle per-epoch: naturally through torch's DataLoader, since `TypedDataset` is a map-style dataset
8. mix-and-matchable multi-step preprocessing: Through parent and child DataSources. Parents define a transformation on their children, which themselves can either be leave datasources or transformations.
9. Ability to avoid materializing the whole dataset in memory at the same time: the underlying hf dataset handles this. Using this setup, a user could theoretically have other kinds of iterables which don't get naturally memory-mapped, but that is an edge case, and the tools are available for the user to handle that themselves in that case.
10. Works with jsonargparse configs: parent-and-child object construction is a natural fit for jsonargparse
11. Type safe as much as possible: `TypedDataset` makes things *much* better than they are now.
12. Happy path is easy: Simple leaf datasource setup is pretty similar to `MirrorDataset` right now, the `format` function can be lifted right out of our current preprocessors, and the collate function can be shared among many models.
13. Multiple datasets can be combined into one (such as for the mixed dataset): easily, through a parent DataSource

## User Responsibilities

This setup requires the user to be careful about a few things:
- Serialization: hf's serialization is nice, but it does have edge cases. This setup requires the user to be aware of those edge cases and handle them appropriately when they come up. Luckily, both the happy path is super easy (just use standard hf processing methods and you get automatic caching), and for edge cases the serialization is completely controlled by the user, so they can build manual solutions if needed.
  - Probably the most dangerous aspect of this proposal is the fact that huggingface can silently fail (or even succeed, then fail on future runs) because of some of this caching behavior. The review process will probably help with this. Also, these problems exist in the current setup anyway.
  - Managing data formats is an unintuitive challenge here. Basically, to work well with hf, everything needs to stay in raw python format (lists, etc) right up until the dataloader. The `collate` function is the one that needs to turn things into torch format.
- Scan avoidance: there is a failure mode where the user does a map or something early on before a take, which could cause a full scan of the dataset up front. This is easily avoided by just doing the take first, but the user must be aware of that.

## Other implementation notes

We'd keep the ability to override a model-provided preprocessor, but instead it would point to a function with the type `DataSource -> DataSource`.

### Why not `IterableDataset`

There was some discussion about using `IterableDataset` over the map style datasets. We were mostly considering this because it seemed to be a way around the scanning problem, and we also thought they'd be faster. The latter turns out not to be true in our use case, and the former is solved easily enough by pre-transformation takes.

## Suggested tickets

1. Implement `TypedDataset`. I actually already have a version of this on another repo. Might be worth factoring out into a small pip package.
2. Change existing `MirrorDataset`s to use preprocessed `TypedDataset`s as the backing dataset. `to_row_type` becomes a (temporary) no-op in all of them, and `take`s are done early to prevent full scans.
3. `MirrorPreprocessor` is removed. Its functionality is moved to `format` and `collate` functions on the various models. Functions that use `preprocess_row` temporarily use `format` wrapped so that it handles one row at a time.
4. `preprocess`, `OnDemandPreprocessedDataset` and their toggles are removed. `format` and `collate` are used directly where needed
5. `MirrorDataset` is fully replaced by `DataSource`.

**Comment by morrisoncheney** [comment]
Got it!

**Comment by jaded0** [comment]
Let's be sure to have assertions and warnings and whatnot so that the risks you pose aren't invisible to us when it comes time to use it. Like, a warning about scan avoidance. A message about location of cache.

We also might be careful not to pair dataset format too much to model if we want it trained to work with a variety of shapes of data. 

The design looks perfectly good and comprehensive.
